### PR TITLE
Add update_platform method to EchoData

### PR DESCRIPTION
**Pull request by imranmaj**
Part of #201 

Adds an `EchoData.update_platform` method which can be used to update the `EchoData.platform` group with additional external platform data.

Partially adapted from https://github.com/ngkavin/echopype/pull/2, https://github.com/ngkavin/echopype/pull/3

### CODE REVIEW

**Comment by leewujung** [comment]
@emiliom : seems like this operation is something we should record in the Provenance group also?

**Comment by emiliom** [comment]
> seems like this operation is something we should record in the Provenance group also?

Great suggestion!! I totally agree.

**Comment by imranmaj** [comment]
@emiliom I was running into issues with trajectory in the saildrone data before so I added that small section to fix it, I can investigate whether it's necessary at all

**Comment by emiliom** [comment]
> @emiliom I was running into issues with trajectory in the saildrone data before so I added that small section to fix it, I can investigate whether it's necessary at all

I think it is necessary. I'm finally getting around to adding the generalized code (done). Now I just need to run the tests.

**Comment by emiliom** [comment]
@imranmaj I'm running into an error with the `~=` operator in `requirements.txt`, again. When doing `conda create -c conda-forge -n echopype_dev --yes python=3.9 --file requirements.txt --file requirements-dev.txt`, I get this error:

```
InvalidVersionSpec: Invalid version '~3.10': invalid operator
```

I think the glitch is different from what I experienced earlier, a couple of weeks ago, but I don't remember the details of that instance. For now, I just need to build a conda env to run the echopype tests, so I changed `~=` to `=`. But this is starting to sound like a persistent problem. If no one else is running into this type of problem, could there be something about my local setup that needs to be updated or changed?

**Comment by leewujung** [comment]
Hmm... I did not run into the ~= problem while creating a new environment with just python=3.8 and echopype. I used conda-forge as usual. @emiliom is the env you are working with a brand new one?

**Comment by emiliom** [comment]
> is the env you are working with a brand new one?

Yeah. I'm creating a new env

File: echopype/echodata/echodata.py [hunk 1]
```diff
@@ -312,6 +312,126 @@ def squeeze_non_scalar(n):
...
+        --------
+        >>> ed = echopype.open_raw(raw_file, "EK60")
+        >>> extra_platform_data = xr.open_dataset(extra_platform_data_file)
+        >>> ed.update_platform(extra_platform_data)
+        """
+
+        # only take data during ping times
+        start_time, end_time = min(self.beam["ping_time"]), max(self.beam["ping_time"])
+
+        # saildrone specific workaround
```
[the remark this answers is not in the document]
Comment (emiliom) [reply echopype/echodata/echodata.py]: I've pushed into your PR the code that generalizes the handling of "trajectory" data encoded using the convention used by the Saildrone sample file

File: echopype/echodata/echodata.py [hunk 2]
```diff
@@ -433,7 +440,7 @@ def mapping_get_multiple(mapping, keys, default=None):
                     mapping_get_multiple(
                         extra_platform_data,
                         ["water_level", "WATER_LEVEL"],
-                        np.ones(num_obs),
+                        default=np.ones(num_obs),
```
Comment (leewujung) [inline echopype/echodata/echodata.py:448]: @emiliom : Why is this `np.ones` instead of `np.zeros` or  use `np.nan` like in other variables?  It seems bizarre to default water level to be 1 m?

File: echopype/echodata/echodata.py [hunk 3]
```diff
@@ -312,6 +312,145 @@ def squeeze_non_scalar(n):
...
+        Data is extracted from `extra_platform_data` by variable name; only the data
+        in `extra_platform_data` with the following variable names will be used:
+            - `"pitch"`
+            - `"roll"`
+            - `"heave"`
+            - `"latitude"`
+            - `"longitude"`
+            - `"water_level"`
+
+        Only `sonar_model`s of `"EK60"`, `"EK80"`, `"EA640"`, and `"AZFP"` are currently supported.
```
Comment (leewujung) [inline echopype/echodata/echodata.py:331]: @imranmaj @emiliom : Could you please remind me why we don't allow this for AD2CP files? I seem to remember that we discussed this but can't recall what the conversation was.

File: echopype/echodata/echodata.py (hunk 2 above)
Comment (emiliom) [reply echopype/echodata/echodata.py:448]: I don't know. I didn't pay attention to that at all. The change I added in my PR was only stylistic -- in all the similar calls to `mapping_get_multiple`, the `default` parameter is used explicitly rather than implicitly (positionally).

File: echopype/echodata/echodata.py (hunk 3 above)
Comment (emiliom) [reply echopype/echodata/echodata.py:331]: I thinks that's for you and @imranmaj to answer. I don't remember being in a conversation about it, and wouldn't know if there's anything specific about AD2CP files that creates a special issue. I assume there isn't, and that its absence was just inertia.

File: echopype/echodata/echodata.py (hunk 2 above)
Comment (leewujung) [reply echopype/echodata/echodata.py:448]: Ah I see, I somehow read this as adding a line, but you added the `default`.
My question should be for @imranmaj  :)

File: echopype/echodata/echodata.py (hunk 3 above)
Comment (leewujung) [reply echopype/echodata/echodata.py:331]: Inertia! 😅 

@imranmaj : do you see problems with adding AD2CP support in this? The potential "problem" I can see is that AD2CP files comes with `pitch`, `roll` and `heading` but not `lat` `lon`, so it is a partially filled dataset. So suppose users then want to add `lat` `lon`, the operations would become reading in the existing Platform group and add the new variables. Is this accurate?

Perhaps more generally the approach would be to:
- check if there is duplications in variables in `extra_platform_data` with the existing `ed.platform` dataset, warn user of overwritting if that's what they select
- add an option to allow user to specify if they want to overwrite or use existing variables
  - if allow overwriting then replace everything
- add new variables not already exist in `ed.platform`

What do you think?

Comment (emiliom) [reply echopype/echodata/echodata.py:331]: For what it's worth, the need you've outlined is more or less identical to the situation for the OOI AZFP data. The platform group is created with some variables, but empty lat & lon.

File: echopype/echodata/echodata.py (hunk 2 above)
Comment (imranmaj) [reply echopype/echodata/echodata.py:448]: I remember talking about this in a meeting, but I don't remember the exact reasoning; I made a PR specifically to make this change https://github.com/ngkavin/echopype/pull/3/files#diff-d35f9ea311a9dd316616bdd6291eb68fdb343808ce6ad80dc6b17402ec56db9aR563

Comment (imranmaj) [reply echopype/echodata/echodata.py:448]: It was np.nan originally

File: echopype/echodata/echodata.py (hunk 3 above)
Comment (leewujung) [reply echopype/echodata/echodata.py:331]: Agreed!

File: echopype/echodata/echodata.py (hunk 2 above)
Comment (leewujung) [reply echopype/echodata/echodata.py:448]: I wonder whether it was because in later operations, say for shipboard deployment, the true depth would be `water_level+range`, and if we have `np.nan` the outcomes would be NaN. I think though it should be `np.zeros` instead of `np.ones`

**Comment by imranmaj** [comment]
External data is being indexed by a dimension named `"time2"` now based on 9/16/21 discussion:

* The sonar netCDF convention says that time dimensions in the Platform group should be named time1, time2, etc
* Indexing using a new time dimension will prevent existing time dimensions (such as location_time) from being overwritten, which in turn will prevent other variables indexed by the previous values of those time dimensions from being invalidated

File: echopype/echodata/echodata.py [hunk 4]
```diff
@@ -312,6 +312,145 @@ def squeeze_non_scalar(n):
...
+            errors="ignore",
+        )
+
+        num_obs = len(extra_platform_data[time_dim])
+
+        def mapping_get_multiple(mapping, keys, default=None):
+            for key in keys:
+                if key in mapping:
+                    return mapping[key].data
+            return default
```
Comment (leewujung) [inline echopype/echodata/echodata.py:399]: @imranmaj: this function would only return the first existing keys in the mapping data set, right?

Seems that this could potentially produce bugs down the line, because the input argument is call `keys`, so users may think that a list with each element being an array of values corresponding to **all** of what's listed in `keys` would be generated, while in fact only the first existing key will be returned.

The current behavior of this function does exactly what you need this function to do in the code below, but in that case, the function name should not be `mapping_get_multiple`, because it would only get data from a single variable.

Which way do you intend this to be? A function that only returns data from 1 variable from all of the possible names, or a function that would return data from multiple variables?

File: echopype/echodata/echodata.py [hunk 5]
```diff
@@ -312,6 +312,145 @@ def squeeze_non_scalar(n):
...
+        >>> extra_platform_data = xr.open_dataset(extra_platform_data_file)
+        >>> ed.update_platform(extra_platform_data)
+        """
+
+        # # only take data during ping times
+        # start_time, end_time = min(self.beam["ping_time"]), max(self.beam["ping_time"])
+
+        # Handle data stored as a CF Trajectory Discrete Sampling Geometry
+        # http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#trajectory-data
+        # The Saildrone sample data file follows this convention
```
Comment (leewujung) [inline echopype/echodata/echodata.py:353]: @emiliom: reading this section you pointed out makes me wonder whether or not it is worthwhile to point out that the lat/lon data in the SONAR-netCDF4 ver.1 is consistent with the H.4.2. Single trajectory specification. Not that this necessitates any actual changes, but there seems to be values to make this explicit.

Comment (leewujung) [reply echopype/echodata/echodata.py:353]: Also, I think once we get around to put this function into the documentation, we should link this explicitly!

File: echopype/echodata/echodata.py [hunk 6]
```diff
@@ -312,6 +312,145 @@ def squeeze_non_scalar(n):
...
+            "long_name"
+        ] = "time dimension for external platform data"
+
+        dropped_vars = []
+        for var in ["pitch", "roll", "heave", "latitude", "longitude", "water_level"]:
+            if var in platform and platform[var].isnull().all():
+                dropped_vars.append(var)
+        if len(dropped_vars) > 0:
+            warnings.warn(
+                f"some variables will be overwritten by platform data: {', '.join(dropped_vars)}"
```
Comment (leewujung) [inline echopype/echodata/echodata.py:386]: Suggesting minor wording changes to clarify which data variables will be overwritten:
`f"some variables in the original Platform group will be overwritten: ...`

File: echopype/echodata/echodata.py (hunk 4 above)
Comment (imranmaj) [reply echopype/echodata/echodata.py:399]: > users may think that a list with each element being an array of values corresponding to all of what's listed in keys would be generated, while in fact only the first existing key will be returned.

It should only be used internally, I don't believe users can use it.

> Which way do you intend this to be? A function that only returns data from 1 variable from all of the possible names, or a function that would return data from multiple variables?

It should return data from 1 variable, it's "searching" for multiple possible names to be flexible.

> the function name should not be `mapping_get_multiple`, because it would only get data from a single variable.

How about `mapping_try_get_multiple`?

File: echopype/echodata/echodata.py [hunk 7]
```diff
@@ -312,6 +312,145 @@ def squeeze_non_scalar(n):
...
+            time2=extra_platform_data[time_dim].values
+        )
+        platform["time2"].attrs[
+            "long_name"
+        ] = "time dimension for external platform data"
+
+        dropped_vars = []
+        for var in ["pitch", "roll", "heave", "latitude", "longitude", "water_level"]:
+            if var in platform and platform[var].isnull().all():
+                dropped_vars.append(var)
```
Comment (leewujung) [inline echopype/echodata/echodata.py:383]: The second part of the logic seems reversed? 
With `platform[var].isnull().all()`, variables with actual non-NaN values in it will return `False` and won't be included in `dropped_vars`.

In the current test, all 6 variables listed contains a single NaN and were included in the dropped list and then replaced. The replacement part is the intended behavior though.

File: echopype/echodata/echodata.py (hunk 4 above)
Comment (leewujung) [reply echopype/echodata/echodata.py:399]: > It should only be used internally, I don't believe users can use it.

I should rephrase this to say that it is not obvious to _others_.

> How about mapping_try_get_multiple?

I think the "multiple" part could be confused with multiple variables. How about `mapping_search_variable`?

**Comment by leewujung** [comment]
Thanks for the changes @imranmaj !  We can now merge this 🙂

File: echopype/echodata/echodata.py (hunk 5 above)
Comment (emiliom) [reply echopype/echodata/echodata.py:353]: > @emiliom: reading this section you pointed out makes me wonder whether or not it is worthwhile to point out that the lat/lon data in the SONAR-netCDF4 ver.1 is consistent with the H.4.2. Single trajectory specification. Not that this necessitates any actual changes, but there seems to be values to make this explicit.

Late reply ... Interesting observation. I agree, the structure is consistent. Maybe the right place to add this observation is the SONAR-netCDF4 version 2 draft text. From what Gavin said, it doesn't look like there are changes in the Platform group in v.2. 

> Also, I think once we get around to put this function into the documentation, we should link this explicitly!

Definitely.

Comment (gavinmacaulay) [reply echopype/echodata/echodata.py:353]: The consistency with the CF single trajectory specification was deliberate :) We tried to use CF conventions whenever it was sensible (and did attempt to meet CF-1.7, although it wasn't feasible for all of sonar-netcdf4 v1).

Comment (emiliom) [reply echopype/echodata/echodata.py:353]: @gavinmacaulay thanks for taking the time to chime in! I've gone ahead and created a new issue in the SONAR-netCDF4 repo. I'll try to handle it myself, hopefully next week.
